### PR TITLE
Makes shieldwall generators actually work and not drain the CPU

### LINKED ISSUE #18578: Shield generators now make shields through walls
**Issue Description**:

Shield generators make shields through walls now. They did not do that before

**What did you expect to happen**: 

Not to make shields through walls

**What happened instead**:

image

**Why is this bad/What are the consequences**:

Breaks some ruins in strange ways / doesn't make sense

**Steps to reproduce the problem**:

Place shield generators with walls between them.
Turn them on.

**When did the problem start happening**:

32203 is the first time I saw it.

### PR DESCRIPTION
**Pull request by farie82**
## What Does This PR Do
Fixes a bug introduced by the appearance rework. Stupid I didn't see it in my review :(
Makes shieldwallgen's not constantly remake the shields. Flikkering the lights doing so and sucking the CPU dry.
Makes the shieldwall's GC proof. If a shieldwallgen gets exploded, thus also killing the shields, it might fail to GC since the shields still hold a reference. I've not verified this though, bit late ATM :).

Also fixes their sprites not updating when they shut down without a player doing it.

Fixes #18578

```
Proc Name                                              Self CPU   Total CPU  Real Time  Overtime  Calls
old
/obj/machinery/shieldwallgen/process                   0.000      0.132      0.136      0.000     248
new
/obj/machinery/shieldwallgen/process                   0.000      0.001      0.001      0.000     248
/obj/machinery/shieldwallgen/proc/activate             0.000      0.000      0.000      0.000     2
/obj/machinery/shieldwallgen/proc/try_link_generators  0.000      0.000      0.376      0.000     8
```
Activate was called twice here to activate the generators. The realtime is due to the sleep which emulates the old behaviour

## Why It's Good For The Game
Fixes performance by about factor 100 and makes them actually usable again.

## Testing
image

Tested by placing them on those spots. Turning them both on and off in random orders. Cutting power beneath them and by cutting power to the station.

For the performance test I turned both on. Then reset the profiler and waited till the `RunQueue` proc hit about 2500 calls.

## Changelog
:cl:
fix: Shieldwall generators now work again
fix: Shieldwall generators now start as expected when they are powered
fix: Shieldwall generators now won't make shields through walls
/:cl:

### CODE REVIEW

File: code/game/machinery/shieldgen.dm [hunk 1]
```diff
@@ -305,142 +305,113 @@
...
+		if(other_generator)
 			break
+		traveled_turfs += current_turf
 
-	if(isnull(G))
+	if(!other_generator?.activated)
 		return
 
-	T2 = loc
+	sleep(rand(1, 4)) // TODO check if we want this behaviour
```
Comment (farie82) [inline code/game/machinery/shieldgen.dm:397]: I'm fine with removing this one if the design team doesn't mind that there is no delay on starting up the generators

File: code/modules/power/singularity/singularity.dm [hunk 2]
```diff
@@ -371,7 +371,7 @@
 			return 0
 	else if(locate(/obj/machinery/shieldwallgen) in T)
 		var/obj/machinery/shieldwallgen/S = locate(/obj/machinery/shieldwallgen) in T
-		if(S && S.active)
+		if(S && S.activated) // Check this
```
Comment (farie82) [inline code/modules/power/singularity/singularity.dm:374]:
```suggestion
		if(S && S.activated)
```
It's fine. I missed line 366

Comment (AffectedArc07) [reply code/modules/power/singularity/singularity.dm]: Cant you also do `if(S?.activated)` here too?

Comment (farie82) [reply code/modules/power/singularity/singularity.dm]: Yep totally can here. I'll change it once I can

File: code/modules/power/singularity/singularity.dm [hunk 3]
```diff
@@ -371,7 +371,7 @@
 			return 0
 	else if(locate(/obj/machinery/shieldwallgen) in T)
 		var/obj/machinery/shieldwallgen/S = locate(/obj/machinery/shieldwallgen) in T
-		if(S && S.active)
+		if(S && S.activated)
```
Comment (farie82) [inline code/modules/power/singularity/singularity.dm:374]:
```suggestion
		if(S?.activated)
```

File: code/game/machinery/shieldgen.dm [hunk 4]
```diff
@@ -305,142 +305,115 @@
...
+	icon_state = "Shield_Gen[activated ? " +a" : ""]"
 
-/obj/machinery/shieldwallgen/proc/power()
-	if(!anchored)
-		power = FALSE
-		return
+/obj/machinery/shieldwallgen/proc/try_charge_shields_power()
 	var/turf/T = loc
 
 	var/obj/structure/cable/C = T.get_cable_node()
```
Comment (hal9000PR) [inline code/game/machinery/shieldgen.dm:327]: can these fit in lockers and stuff? could cause issues if their loc isnt a turf.

File: code/game/machinery/shieldgen.dm [hunk 5]
```diff
@@ -305,142 +305,115 @@
...
-			setup_field(2)
-		spawn(3)
-			setup_field(4)
-		spawn(4)
-			setup_field(8)
-	if(active)
-		if(!power)
-			visible_message("<span class='warning'>[name] shuts down due to lack of power!</span>", \
+	if(!activated)
+		return
```
Comment (hal9000PR) [inline code/game/machinery/shieldgen.dm:370]: process kill?

File: code/game/machinery/shieldgen.dm [hunk 6]
```diff
@@ -305,142 +305,115 @@
...
+	if(!try_charge_shields_power())
+		visible_message("<span class='warning'>[name] shuts down due to lack of power!</span>", \
 				"You hear heavy droning fade out")
```
Comment (hal9000PR) [inline code/game/machinery/shieldgen.dm:373]: I think the second arg is the message sent to the src. when this looks like it wants to be a blind message?
might want to use a named arg here.

File: code/game/machinery/shieldgen.dm [hunk 7]
```diff
@@ -305,142 +305,115 @@
...
-	for(var/dist = 0, dist < steps, dist += 1) // creates each field tile
-		var/field_dir = get_dir(T2,get_step(T2, NSEW))
-		T = get_step(T2, NSEW)
-		T2 = T
-		var/obj/machinery/shieldwall/CF = new/obj/machinery/shieldwall/(src, G) //(ref to this gen, ref to connected gen)
-		CF.loc = T
-		CF.dir = field_dir
+	var/opposite_direction = turn(direction, 180)
+	for(var/T in traveled_turfs)
+		var/obj/machinery/shieldwall/SW = new/obj/machinery/shieldwall(T, src, other_generator) //(ref to this gen, ref to connected gen)
```
Comment (hal9000PR) [inline code/game/machinery/shieldgen.dm:403]:
```suggestion
		var/obj/machinery/shieldwall/SW = new /obj/machinery/shieldwall(T, src, other_generator) //(ref to this gen, ref to connected gen)
```

**Review by hal9000PR — CHANGES_REQUESTED** [review]
looks mostly good. my only question is do these have to be processing constantly? couldn't you make it only process when you turn them on/off?

File: code/game/machinery/shieldgen.dm (hunk 6 above)
Comment (farie82) [reply code/game/machinery/shieldgen.dm:373]: There are two versions of `visible_message`. One on mobs and one on everything else. The mobs one will take a `self_message` param on the second place. The base (atom) one will only take the message and the blind message.

File: code/game/machinery/shieldgen.dm (hunk 4 above)
Comment (farie82) [reply code/game/machinery/shieldgen.dm:327]: They should never be located in a locker or such. If this happens then something got fucked up majorly and it should cause a runtime

File: code/game/machinery/shieldgen.dm (hunk 5 above)
Comment (farie82) [reply code/game/machinery/shieldgen.dm]: I've ensured the shieldgen won't process when it ain't active

**Review by hal9000PR — APPROVED** [review]

**Pull request merged**